pkg/controller/performanceprofile/performanceprofile_controller.go

```diff
@@ -103,54 +208,225 @@ func (r *ReconcilePerformanceProfile) Reconcile(request reconcile.Request) (reco
+			removeFinalizer(instance, finalizer)
+			if err := r.client.Update(context.TODO(), instance); err != nil {
+				return reconcile.Result{}, err
+			}
+
+			return reconcile.Result{}, nil
+		}
+	}
+
+	// TODO: we need to check if all under performance profiles values != nil
+	// first we need to decide if each of values required and we should move the check into validation webhook
+	// for now let's assume that all parameters needed for assets scrips are required
+	if err := r.verifyPerformanceProfileParameters(instance); err != nil {
+		// we do not want to reconcile again in case of error, because a user will need to update the PerformanceProfile anyway
+		klog.Errorf("failed to reconcile: %v", err)
+		return reconcile.Result{}, nil
 	}
 
-	// Check if this Pod already exists
-	found := &corev1.Pod{}
-	err = r.client.Get(context.TODO(), types.NamespacedName{Name: pod.Name, Namespace: pod.Namespace}, found)
-	if err != nil && errors.IsNotFound(err) {
-		reqLogger.Info("Creating a new Pod", "Pod.Namespace", pod.Namespace, "Pod.Name", pod.Name)
-		err = r.client.Create(context.TODO(), pod)
-		if err != nil {
+	// add finalizer
+	if !hasFinalizer(instance, finalizer) {
+		instance.Finalizers = append(instance.Finalizers, finalizer)
+		if err := r.client.Update(context.TODO(), instance); err != nil {
 			return reconcile.Result{}, err
 		}
 
-		// Pod created successfully - don't requeue
+		// we exit reconcile loop because we will have additional update reconcile
 		return reconcile.Result{}, nil
-	} else if err != nil {
+	}
+
+	// apply components
+	if err := r.applyComponents(instance); err != nil {
+		klog.Errorf("failed to deploy components: %v", err)
 		return reconcile.Result{}, err
 	}
 
-	// Pod already exists - don't requeue
-	reqLogger.Info("Skip reconcile: Pod already exists", "Pod.Namespace", found.Namespace, "Pod.Name", found.Name)
+	// TODO: we need to update the status
+
 	return reconcile.Result{}, nil
 }
 
-// newPodForCR returns a busybox pod with the same name/namespace as the cr
-func newPodForCR(cr *performancev1alpha1.PerformanceProfile) *corev1.Pod {
-	labels := map[string]string{
-		"app": cr.Name,
+func (r *ReconcilePerformanceProfile) applyComponents(profle *performancev1alpha1.PerformanceProfile) error {
+	// deploy machine config pool
+	mcp := machineconfigpool.NewPerformance(profle)
+	if err := controllerutil.SetControllerReference(profle, mcp, r.scheme); err != nil {
+		return err
 	}
-	return &corev1.Pod{
-		ObjectMeta: metav1.ObjectMeta{
-			Name:      cr.Name + "-pod",
-			Namespace: cr.Namespace,
-			Labels:    labels,
-		},
-		Spec: corev1.PodSpec{
-			Containers: []corev1.Container{
-				{
-					Name:    "busybox",
-					Image:   "busybox",
-					Command: []string{"sleep", "3600"},
-				},
-			},
-		},
+	if err := r.createOrUpdateMachineConfigPool(mcp); err != nil {
+		return err
+	}
+
+	// pause machine config pool
+	if err := r.pauseMachineConfigPool(mcp.Name, true); err != nil {
```

Comment:
we don't want to issu APIs call every time the reconcile loop pops unless something changes.  We'll need to check to see if modifications are going to occur and only pause/unpause as modifications happen.

Also, this pause/unpause logic seems like it could have a race condition. For example, what guarantees us that the machine config operator sees that we paused a mcp before seeing a new machine config is posted?

Reply:
What do you mean by modifications? Modifications to performance profile resource? We anyway run a very small number of reconciling loops because of our predicates(only when spec or labels was updated, so we really pause only on important modifications).

Regarding the race condition, you are right once we will have `MaxConcurrentReconciles` bigger than one we will have troubles, but to be honest I do not have a good solution for it.
I can lock with the mutex `applyComponents` method, but it the same as running reconcile only with the one thread, and you can not really know if it is an additional update in progress.
For now, I will add a lock, but maybe do you have better ideas?

Reply:
> What do you mean by modifications? Modifications to performance profile resource? We anyway run a very small number of reconciling loops because of our predicates(only when spec or labels was updated, so we really pause only on important modifications).

What I'm suggesting is to determine up front whether any config changes will occur that involve a restart, and only perform the "pause/unpause" logic when it's detected that changes will occur.

so the flow would be something like this.

* look though cache to determine if any machine configs, kubelet configs, etc... either need to be posted or updated.
    * If not, fast succeed and move on to status calculations since there's nothing to be done
    * else pause, apply changes, unpause then move on to status calculations.

I know what I'm suggesting is kind of a pain, the issue is if we get into an error loop or some other unexpected reconcile loop spin, things like this mean we spam the api server with writes.

In practice we need to only GET or POST to the api server when it's absolutely necessary. This is why we do things like only posting an update to the status section when ```if ! reflect.DeepEquals(newStatus, oldStatus)``` We're avoiding a post to the API server that would otherwise occur every reconcile loop if we didn't determine that the change wasn't necessary by looking at cache.

Reply:
> Regarding the race condition, you are right once we will have MaxConcurrentReconciles bigger than one we will have troubles, but to be honest I do not have a good solution for it.
I can lock with the mutex applyComponents method, but it the same as running reconcile only with the one thread, and you can not really know if it is an additional update in progress.
For now, I will add a lock, but maybe do you have better ideas?

And with regards to the race condition.

What I was referring to isn't related to our reconcile loop. Regardless of how many concurrent reconcile loops we allow, we are guaranteed a single key is only being acted on by a single reconcile execution. There's no parallel execution of the same key, only parallel execution of separate keys. 

What I was referring to is actually the machine config operator's reconcile loop. Is there anything guaranteeing us that the operator will be notified that we paused a pool before it's notified we posted new machine configs or kubelet configs?

Basically, do we have any guarantees that the order we post changes in our reconcile loop matches the order another operator will receive them in? so, If I post resource A then resource B, is it possible for another operator to observe resource B's creation before resource A? if resources A and B are of different kinds, i think this might actually be possible.

Reply:
> so the flow would be something like this.
> 
> * look though cache to determine if any machine configs, kubelet configs, etc... either need to be posted or updated.
>   
>   * If not, fast succeed and move on to status calculations since there's nothing to be done
>   * else pause, apply changes, unpause then move on to status calculations.

@davidvossel But we already have this functionality because of custom update predicates, it just does not enter to the reconcile loop at all, if it wasn't important update.

Reply:
> And with regards to the race condition.
> 
> What I was referring to isn't related to our reconcile loop. Regardless of how many concurrent reconcile loops we allow, we are guaranteed a single key is only being acted on by a single reconcile execution. There's no parallel execution of the same key, only parallel execution of separate keys.
> 
> What I was referring to is actually the machine config operator's reconcile loop. Is there anything guaranteeing us that the operator will be notified that we paused a pool before it's notified we posted new machine configs or kubelet configs?
> 
> Basically, do we have any guarantees that the order we post changes in our reconcile loop matches the order another operator will receive them in? so, If I post resource A then resource B, is it possible for another operator to observe resource B's creation before resource A? if resources A and B are of different kinds, i think this might actually be possible.

Ah, I see, good point, but I unsure how we can avoid it, I can divide reconcile loop, something like:
- create or update MCP and pause it(`return reconcile.Result{Requeue: true, RequeueAfter: time.Second}`)
- before creating all other components, check that MCP paused, if return again `reconcile.Result{Requeue: true, RequeueAfter: 10*time.Second}`
- create all components
- unpause MCP
@davidvossel WDT?

Reply:
Looks good, the only thing I'd add is to only unpause after you observe all dependent resources have been created.

Reply:
Hm, we unpause just when all mutated objects equal to nil https://github.com/openshift-kni/performance-addon-operators/blob/c3dd292b406b8ae95b4047ea8c5f12661c159b64/pkg/controller/performanceprofile/performanceprofile_controller.go#L366, it means that the object that we get from the cache and the object that we expect to have it the same, so it means that all objects already exist, or do I miss something?